administrator/components/com_templates/src/Model/TemplateModel.php

```diff
@@ -1985,4 +2026,230 @@ protected function checkFormat($ext)
+
+				if (!$result) {
+					$app->enqueueMessage(Text::_('COM_TEMPLATES_ERROR_COULD_NOT_WRITE'), 'error');
+
+					return false;
+				}
+			}
+
+			return true;
+		} else {
+			$app->enqueueMessage(Text::_('COM_TEMPLATES_ERROR_INVALID_FROM_NAME'), 'error');
+
+			return false;
+		}
+	}
+
+	/**
+	 * Method to get the folder structure as an object
+	 *
+	 * @param   string   $dir           The base directory
+	 * @param   boolean  $relativePath  Use relative path
+	 *
+	 * @return  object
+	 *
+	 * @since	__DEPLOY_VERSION__
+	 */
+	protected function getDirContents($dir, $relativePath = false)
+	{
+		$fileList = [];
+
+		$iterator = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($dir));
+
+		foreach ($iterator as $file) {
+			if ($file->isDir() && !in_array($file->getPath(), ['.', '..'])) {
+				if ($relativePath) {
+					$path = str_replace($dir, '', $file->getPath());
+					$path = ltrim($path, '/\\');
+				}
+
+				if ($path !== '') {
+					$shouldAdd = true;
+
+					foreach ($fileList as $obj)
+					{
+						if ($obj->name === $path)
+						{
+							$shouldAdd = false;
+							break;
+						}
+					}
+
+					if ($shouldAdd)
+					{
+						$temp       = new \stdClass;
+						$temp->type = 'folder';
+						$temp->name = basename($path);
+						$temp->path = $path;
+						$fileList[] = $temp;
+					}
+				}
+
+				continue;
+			}
+
+			$path = $file->getPathname();
+
+			if ($relativePath) {
+				$path = str_replace($dir, '', $path);
+				$path = ltrim($path, '/\\');
+			}
+
+			$temp         = new \stdClass;
+			$temp->type   = 'file';
+			$temp->name   = basename($path);
+			$temp->folder = dirname($path);
+			$fileList[]   = $temp;
+		}
+
+		return $fileList;
+	}
+
+	/**
+	 * Converts a filesystem tree to a PHP array.
+	 *
+	 * @param   string  $dir  The base directory
+	 *
+	 * @return  object
+	 *
+	 * @since	__DEPLOY_VERSION__
+	 */
+	protected function dirToArray($dir)
+	{
+		if (!is_dir($dir))
+		{
+			// If the user supplies a wrong path we inform him.
+			return null;
+		}
+
+		// Our PHP representation of the filesystem
+		// for the supplied directory and its descendant.
+		$data = [];
+
+		foreach (new \DirectoryIterator($dir) as $f)
+		{
+			if ($f->isDot())
+			{
+				// Dot files like '.' and '..' must be skipped.
+				continue;
+			}
+
+			$path = $f->getPathname();
+			$name = $f->getFilename();
+
+			if ($f->isFile() && $name !== '.DS_Store')
+			{
+				$data[] = [
+					'file' => $name,
+					'path' => str_replace(JPATH_ROOT, '', $path),
+				];
+			}
+			else
+			{
+				if (in_array(basename($path), ['node_modules']))
+				{
+					continue;
+				}
+
+				// Process the content of the directory.
+				$files = $this->dirToArray($path);
+
+				$data[] = [
+					'dir'  => $files,
+					'name' => $name,
+					'path' => str_replace(JPATH_ROOT, '', $path),
+				];
+				// A directory has a 'name' attribute
+				// to be able to retrieve its name.
+				// In case it is not needed, just delete it.
```

Comment:
```suggestion
				// A directory has a 'name' attribute  to be able to retrieve its name.
				// In case it is not needed, just delete it.
```

only two line comments allowed with //